js/test/doughnut.spec.js

```diff
@@ -1,50 +1,140 @@
+    184, 118, 189, 149, 226, 203, 63, 146, 23, 217, 177, 9]);
+
+const expiry = 987654321;
+
+const notBefore = 12345;
+
+const defaultSignatureBeforeSigning = Uint8Array.from({length: 64}, x => 0);
+
+const encodedDoughnut = new Uint8Array(
+    composeDoughnut({
+        holder,
+        issuer: ed25519Keypair.publicKey,
+        signature,
+    })
+);
+
+describe('wasm doughnut', () => {
+    describe('Decoded instance', () => {
+        test('getters work', () => {
+            let d = Doughnut.decode(encodedDoughnut);
+            // Fields are correct
+            expect(d.holder()).toEqual(holder);
+            expect(d.issuer()).toEqual(ed25519Keypair.publicKey);
+            expect(d.expiry()).toEqual(expiry);
+            expect(d.notBefore()).toEqual(notBefore);
+            expect(d.payloadVersion()).toEqual(0);
+            expect(d.signatureVersion()).toEqual(1);
+            expect(d.signature()).toEqual(signature);
+
+            // encodes the same
+            expect(d.encode()).toEqual(encodedDoughnut);
+        });
+    });
+
+    describe('Class instance', () => {
+        test('getters work', () => {
+            const d = new Doughnut(
+                ed25519Keypair.publicKey,
+                holder,
+                expiry,
+                notBefore
+            );
+
+            expect(d.holder()).toEqual(holder);
+            expect(d.issuer()).toEqual(ed25519Keypair.publicKey);
+            expect(d.expiry()).toEqual(expiry);
+            expect(d.notBefore()).toEqual(notBefore);
+
+            const defaultSignatureVersionBeforeSigning = 0;
+            expect(d.signatureVersion()).toEqual(defaultSignatureVersionBeforeSigning);
+
+            expect(d.payloadVersion()).toEqual(0);
+            expect(d.signature()).toEqual(defaultSignatureBeforeSigning);
+        });
+    });
+
+    describe('Schnorrkel', () => {
+        test('sr25519 signing is verifiable', () => {
+            const d = new Doughnut(
+                sr25519Keypair.publicKey,
+                holder,
+                expiry,
+                notBefore
+            );
+
+            expect(d.signature()).toEqual(defaultSignatureBeforeSigning);
+
+            d.signSr25519(sr25519Keypair.secretKey);
+
+            expect(d.signatureVersion()).toEqual(0);
+
+            expect(d.verify(holder, 12346)).toEqual(true);
+        });
+    });
+
+    describe('ed25519', () => {
+        test('ed25519 signing produce the expected signature', () => {
+            let d = new Doughnut(
+                ed25519Keypair.publicKey,
+                holder,
+                expiry,
+                notBefore
+            );
+
+            expect(d.signature()).toEqual(defaultSignatureBeforeSigning);
+
+            d.signEd25519(ed25519Keypair.secretKey);
+
+            expect(d.signatureVersion()).toEqual(1);
+
+            expect(d.signature()).toEqual(signature);
```

Comment:
need to check `verify()` here also

Reply:
signature version is only set correctly after signing the doughnut. This is our current expected behaviour. Why should the test fail for that? The same as for the signature. When the doughnut is not signed, it returns a zero-filled array and therefore expected. Once we make the signature and signature version as Options then the correct behaviour would change and would affect this test.

Reply:
I added verify to this test as well.